kvserver: honor priority on send and receive snapshots

AlexTalks opened the pull request:
Previously on both the send and receive snapshot side, the
various places that sent snapshots were uncoordinated, so
the choice of which snapshot was sent was somewhat arbitrary.
This PR uses the new multiqueue to prioritize them correctly.

Release note (performance improvement): Snapshots use a fair
round-robin approach for choosing which one to send next. This
allows decommissioning to complete much faster.

Release Justification: A number of customer support cases
are caused by incorrect snapshot prioritization.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/86701)
<!-- Reviewable:end -->

andrewbaptist commented:
Screenshot of a run of `decommissionBench/nodes=8/cpu=16/warehouses=3000/drain-first/while-upreplicating` after this change:

![screencapture-104-196-61-150-3000-d-J-yAVzkVk2-decommissions-full-2022-08-25-17_38_57](https://user-images.githubusercontent.com/46975918/186773880-e09ecb5b-9c88-4323-8225-9559c544ebe3.png)

AlexTalks commented:
This can be rebased onto master now, removing the first two commits (or we can switch to a new branch - either way is fine with me!)

andrewbaptist commented:
Rebased onto master.

andrewbaptist commented:
From a testing perspective, the decommissioning tests were all run. With a combination of this and Lidor's changes, the time to decommission has dropped from ~55 minutes to ~24 minutes. This will be even faster if drain is called before decommission. 
We also tested version compatibility (v22.1 and this branch) and the results were positive (no errors and similar behavior in the mixed cluster to v22.1). 
Finally the moving hotkey (DD) test  https://docs.google.com/document/d/1AMvxf7uXve8rV21DXEeKYx--bswDy2OSSw_aKenG9c8/edit#heading=h.8ssuvtfzpdk7 was run and worked "well" even during a decommission of a node. Ranges continued to move as expected.

nvb reviewed (commented):
You were asking about adding a cluster version flag in https://cockroachlabs.slack.com/archives/C02H8JCC7DM/p1661545679399539. That seems apt to me, as there is a possibility that without such a flag, mixed version clusters will prioritize snapshots from v22.2 nodes. You could avoid that behavior by adding a sender-side cluster version check that strips the source and priority fields before sending a `SnapshotRequest` if the cluster is not yet all running v22.2 processes.

We also talked about adding a cluster setting to disable this. Did you still want to do that? If so, the logic could live in the same place as the cluster version check.

> Reviewed 8 of 8 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-:-NAeLC1a6sosohcAwEyS:b-acoc98)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15 and @AlexTalks)

___
*[`pkg/kv/kvserver/replica_learner_test.go` line 246 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-NAeGXws5NE9h2Leucnb:-NAeGXws5NE9h2Leucnc:b-e5m57i) ([raw file](https://github.com/cockroachdb/cockroach/blob/666ca0102b2d4ca9bfaa05b0505040efb152d5c0/pkg/kv/kvserver/replica_learner_test.go#L246)):*
> ```Go
> 	// This test assumes that the snapshot send limit is set to 1 so the second
> 	// snapshot send will block waiting for the first to complete.
> 	cleanup := envutil.TestSetEnv(t, "COCKROACH_CONCURRENT_SNAPSHOT_SEND_LIMIT", "1")
> ```

The use of env vars to control a TestCluster feels like a bit of an anti-pattern. Can we copy `concurrentSnapshotApplyLimit` and `concurrentSnapshotSendLimit` to `KVConfig` and `TestServerArgs` and then configure it directly in this test? See `ScanMinIdleTime` for an example of how that plumbing would work.
___
*[`pkg/kv/kvserver/store.go` line 1161 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-NAeJ6ZG5rrkrJC_e4nT:-NAeJ6ZG5rrkrJC_e4nU:bvkgzuz) ([raw file](https://github.com/cockroachdb/cockroach/blob/666ca0102b2d4ca9bfaa05b0505040efb152d5c0/pkg/kv/kvserver/store.go#L1161)):*
> ```Go
> 	if sc.concurrentSnapshotSendLimit == 0 {
> 		sc.concurrentSnapshotSendLimit =
> 			envutil.EnvOrDefaultInt("COCKROACH_CONCURRENT_SNAPSHOT_SEND_LIMIT", 2)
> ```

Let's comment here (or `server/config.go` if we do my suggestion from above) about why we want an apply-size limit of 1 but a send-side limit of 2.
___
*[`pkg/kv/kvserver/store_snapshot.go` line 691 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-NAeJ_8h4PxTi7A5ThMO:-NAeJ_8h4PxTi7A5ThMP:bhu7r4l) ([raw file](https://github.com/cockroachdb/cockroach/blob/666ca0102b2d4ca9bfaa05b0505040efb152d5c0/pkg/kv/kvserver/store_snapshot.go#L691)):*
> ```Go
> 		fn()
> 	}
> 	requestSource := int32(req.SenderQueueName)
> ```

nit: there seems to be a good amount of repetition between this function and the previous one. That's not something to solve here, but then let's make the exact same changes to each so it doesn't look like there's a difference. For instance, define these variables in both or in neither.
___
*[`pkg/kv/kvserver/store_snapshot.go` line 715 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-NAeJwly20uYFLv1xx5_:-NAeJwly20uYFLv1xx5a:b717ke3) ([raw file](https://github.com/cockroachdb/cockroach/blob/666ca0102b2d4ca9bfaa05b0505040efb152d5c0/pkg/kv/kvserver/store_snapshot.go#L715)):*
> ```Go
> 	waitingSnapshotMetric, inProgressSnapshotMetric, totalInProgressSnapshotMetric *metric.Gauge,
> ) (cleanup func(), _err error) {
> 	// TODO: Is this necessary
> ```

Yes, it is necessary if you want the caller to be able to call the function on error. However, this contract that the cleanup function must be called even on error is not standard. Usually, functions will handle this themselves on error paths and not hand the responsibility to their caller. I think here that would just mean calling `semaphore.Cancel(task)` directly in the `case <-queueCtx.Done():` case. Consider switching to that approach.
___
*[`pkg/kv/kvserver/store_test.go` line 2929 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-NAeKpcR-yG87ZbB6kdt:-NAeKpcR-yG87ZbB6kdu:b4gkqkn) ([raw file](https://github.com/cockroachdb/cockroach/blob/666ca0102b2d4ca9bfaa05b0505040efb152d5c0/pkg/kv/kvserver/store_test.go#L2929)):*
> ```Go
> }
> 
> func TestSendSnapshotConcurrency(t *testing.T) {
> ```

nit: Could you give this test a comment explaining what it is testing.


<!-- Sent from Reviewable.io -->

andrewbaptist reviewed (commented):
Alex and I went back and forth on the cluster version and enable/disable flags. I added the check for cluster version but did not add an ability to disable the feature. I wanted to avoid an unnecessary extra check here, and it wasn't clear when we would want to disable this feature. Let me know if you think we need a flag to disable this feature also.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-:-NAfNz8FCKKcVRGjOGKx:bigk1bt)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15 and @nvanbenschoten)

___
*[`pkg/kv/kvserver/replica_learner_test.go` line 246 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-NAeGXws5NE9h2Leucnb:-NAf6oQc4aegRc4FzP6_:b81gxh6) ([raw file](https://github.com/cockroachdb/cockroach/blob/666ca0102b2d4ca9bfaa05b0505040efb152d5c0/pkg/kv/kvserver/replica_learner_test.go#L246)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

The use of env vars to control a TestCluster feels like a bit of an anti-pattern. Can we copy `concurrentSnapshotApplyLimit` and `concurrentSnapshotSendLimit` to `KVConfig` and `TestServerArgs` and then configure it directly in this test? See `ScanMinIdleTime` for an example of how that plumbing would work.
</blockquote></details>

Done, This required touching a few more places but is cleaner and statically compiled.
___
*[`pkg/kv/kvserver/store.go` line 1161 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-NAeJ6ZG5rrkrJC_e4nT:-NAeXPW_EQUWJhQHY1MY:b2izebo) ([raw file](https://github.com/cockroachdb/cockroach/blob/666ca0102b2d4ca9bfaa05b0505040efb152d5c0/pkg/kv/kvserver/store.go#L1161)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Let's comment here (or `server/config.go` if we do my suggestion from above) about why we want an apply-size limit of 1 but a send-side limit of 2.
</blockquote></details>

Done, added comment
___
*[`pkg/kv/kvserver/store_snapshot.go` line 691 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-NAeJ_8h4PxTi7A5ThMO:-NAeXYd8DJkk9vRUsJtY:btdlom1) ([raw file](https://github.com/cockroachdb/cockroach/blob/666ca0102b2d4ca9bfaa05b0505040efb152d5c0/pkg/kv/kvserver/store_snapshot.go#L691)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: there seems to be a good amount of repetition between this function and the previous one. That's not something to solve here, but then let's make the exact same changes to each so it doesn't look like there's a difference. For instance, define these variables in both or in neither.
</blockquote></details>

Done, made them consistent (don't declare variable)
___
*[`pkg/kv/kvserver/store_snapshot.go` line 715 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-NAeJwly20uYFLv1xx5_:-NAeXi5g3KZ2ucioTDpf:bng1tg3) ([raw file](https://github.com/cockroachdb/cockroach/blob/666ca0102b2d4ca9bfaa05b0505040efb152d5c0/pkg/kv/kvserver/store_snapshot.go#L715)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Yes, it is necessary if you want the caller to be able to call the function on error. However, this contract that the cleanup function must be called even on error is not standard. Usually, functions will handle this themselves on error paths and not hand the responsibility to their caller. I think here that would just mean calling `semaphore.Cancel(task)` directly in the `case <-queueCtx.Done():` case. Consider switching to that approach.
</blockquote></details>

Done, found a different style here that works much better!
___
*[`pkg/kv/kvserver/store_test.go` line 2929 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-NAeKpcR-yG87ZbB6kdt:-NAf69d48BSdHgsQKo-s:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/666ca0102b2d4ca9bfaa05b0505040efb152d5c0/pkg/kv/kvserver/store_test.go#L2929)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: Could you give this test a comment explaining what it is testing.
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Alex and I went back and forth on the cluster version and enable/disable flags. I added the check for cluster version but did not add an ability to disable the feature. I wanted to avoid an unnecessary extra check here, and it wasn't clear when we would want to disable this feature. Let me know if you think we need a flag to disable this feature also.

Yeah, I do think we should add an enable/disable flag, even if it is private and we plan to remove it in the next release. It's pretty easy to do (we already have the version gate), and being able to disable this prioritization in a customer outage scenario might come in handy. We very rarely regret these kinds of non-exposed knobs when we roll in potentially risky changes, especially late in a release. They're either harmless and we can remove them once we have more confidence or they save the day when a customer gets into a tricky spot.

> Reviewed 7 of 10 files at r2, 6 of 6 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-:-NAjDpcNEbEi_y8YMTFu:b-lp5qj4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @AlexTalks, and @nvanbenschoten)

___
*[`pkg/kv/kvserver/store_snapshot.go` line 693 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-NAjCpbHD01osk_D5Bjh:-NAjCpbHD01osk_D5Bji:bgrj6zo) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff68074c30040bcddfd51f4e1cf20cbf33e42c0a/pkg/kv/kvserver/store_snapshot.go#L693)):*
> ```Go
> 	// TODO(baptist): Remove this block in v23.1 once this version flag goes away
> 	// Don't send a queue name or priority if the receiver may not understand them.
> 	if !s.ClusterSettings().Version.IsActive(ctx, clusterversion.PrioritizeSnapshots) {
> ```

Don't we want this in `Replica.sendSnapshot`, so that the name and priority won't make it to the receiver (who might be a v22.2 node)?
___
*[`pkg/kv/kvserver/store_snapshot.go` line 709 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-NAjDChBAtrV7DBofuU8:-NAjDChBAtrV7DBofuU9:b-jcnl64) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff68074c30040bcddfd51f4e1cf20cbf33e42c0a/pkg/kv/kvserver/store_snapshot.go#L709)):*
> ```Go
> // throttleSnapshot is a helper function to throttle snapshot sending and
> // receiving. The returned closure is used to cleanup the reservation and
> // release its resources and must be called even if there is an error.
> ```

I think this comment is now stale, as are the comments on the preceding two methods.


<!-- Sent from Reviewable.io -->

andrewbaptist reviewed (commented):
Done, added enable/disable flag.

> Reviewed 1 of 10 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-:-NAjuRfX7Eqst62rsH1f:bqedwop)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @andrewbaptist, and @nvanbenschoten)

___
*[`pkg/kv/kvserver/store_snapshot.go` line 693 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-NAjCpbHD01osk_D5Bjh:-NAjw0IpE3Z28_Q84kIu:b-lz1spk) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff68074c30040bcddfd51f4e1cf20cbf33e42c0a/pkg/kv/kvserver/store_snapshot.go#L693)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Don't we want this in `Replica.sendSnapshot`, so that the name and priority won't make it to the receiver (who might be a v22.2 node)?
</blockquote></details>

Done, moved 
___
*[`pkg/kv/kvserver/store_snapshot.go` line 709 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-NAjDChBAtrV7DBofuU8:-NAjmjJT4ZswN-xHD5n4:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff68074c30040bcddfd51f4e1cf20cbf33e42c0a/pkg/kv/kvserver/store_snapshot.go#L709)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I think this comment is now stale, as are the comments on the preceding two methods.
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r5, 2 of 2 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-:-NAkHl-SAgKK74xxbExs:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aayushshah15)



<!-- Sent from Reviewable.io -->

AlexTalks reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86701)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aayushshah15)

___
*[`pkg/kv/kvserver/replica_command.go` line 2671 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-NAko2Hg4SKjS4ie8-HL:-NAko2Hg4SKjS4ie8-HM:bod8ddx) ([raw file](https://github.com/cockroachdb/cockroach/blob/81a270e86c1c64052a1a0e5bab9598a4e83c7021/pkg/kv/kvserver/replica_command.go#L2671)):*
> ```Go
> 	// queue name or priority if the receiver may not understand them or the
> 	// setting is disabled.
> 	if !r.store.ClusterSettings().Version.IsActive(ctx, clusterversion.PrioritizeSnapshots) ||
> ```

nit: Might be worth clarifying that the only thing we *for sure* want to remove in 23.1 is the cluster version.
___
*[`pkg/kv/kvserver/store.go` line 259 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-NAkoQaS3VvlvsLPcH9-:-NAkoQaS3VvlvsLPcH90:b4ouwvc) ([raw file](https://github.com/cockroachdb/cockroach/blob/81a270e86c1c64052a1a0e5bab9598a4e83c7021/pkg/kv/kvserver/store.go#L259)):*
> ```Go
> 		ProtectedTimestampReader:    spanconfig.EmptyProtectedTSReader(clock),
> 		SnapshotSendLimit:           2,
> 		SnapshotApplyLimit:          1,
> ```

I think (as we discussed) this should not use hard-coded values.  Apologies for making you keep moving these around, but maybe it would make sense to move the defaults into `kvserver` and that way you can reference them both here and in the `server` pkg.  You'll just have to make the defaults public however.



_Code quote:_
```Go
		SnapshotSendLimit:           2,
		SnapshotApplyLimit:          1,
```
___
*[`pkg/kv/kvserver/store_snapshot.go` line 67 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-NAkthLnEaWOioiNdHJ4:-NAkthLnEaWOioiNdHJ5:bbgd6cg) ([raw file](https://github.com/cockroachdb/cockroach/blob/81a270e86c1c64052a1a0e5bab9598a4e83c7021/pkg/kv/kvserver/store_snapshot.go#L67)):*
> ```Go
> var snapshotPrioritizationEnabled = settings.RegisterBoolSetting(
> 	settings.SystemOnly,
> 	"kv.snapshot.prioritization_enabled",
> ```

nit: `kv.snapshot_prioritization.enabled`
___
*[`pkg/kv/kvserver/store_snapshot.go` line 69 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-NAktlAh7OCMKyoEssnl:-NAktlAh7OCMKyoEssnm:b-6wzlve) ([raw file](https://github.com/cockroachdb/cockroach/blob/81a270e86c1c64052a1a0e5bab9598a4e83c7021/pkg/kv/kvserver/store_snapshot.go#L69)):*
> ```Go
> 	"kv.snapshot.prioritization_enabled",
> 	"if true, then prioritize enqueued snapshots on both the send or receive sides",
> 	true)
> ```

nit: does this pass lint?  I'd assume we'd need:
```
  true,
)
```
___
*[`pkg/kv/kvserver/store_snapshot.go` line 676 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-NAku1TZ5gr8yDIUv19d:-NAku1TZ5gr8yDIUv19e:b-fc53ix) ([raw file](https://github.com/cockroachdb/cockroach/blob/81a270e86c1c64052a1a0e5bab9598a4e83c7021/pkg/kv/kvserver/store_snapshot.go#L676)):*
> ```Go
> 
> 	return s.throttleSnapshot(ctx, s.snapshotApplySem,
> 		int(header.SenderQueueName), header.SenderQueuePriority,
> ```

nit: I still don't love the casting to `int`s, it feels like we should at least have had a `multiqueue.QueueID` type or something.  This isn't something I'd worry about changing now, but just wanted to note it.
___
*[`pkg/kv/kvserver/store_snapshot.go` line 723 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-NAkuLHm8aSt77dREKAA:-NAkuLHm8aSt77dREKAB:b-i7b16b) ([raw file](https://github.com/cockroachdb/cockroach/blob/81a270e86c1c64052a1a0e5bab9598a4e83c7021/pkg/kv/kvserver/store_snapshot.go#L723)):*
> ```Go
> 	// getting stuck behind large snapshots managed by the replicate queue.
> 	if rangeSize != 0 || s.cfg.TestingKnobs.ThrottleEmptySnapshots {
> 		task := semaphore.Add(requestSource, requestPriority)
> ```

should we still call this `semaphore`?  It seems inaccurate - can we rename to something like `snapshotQueue`?
___
*[`pkg/kv/kvserver/store_test.go` line 2933 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-NAkwNsXCu9MYfc8v2Qo:-NAkwNsXCu9MYfc8v2Qp:btyfc4v) ([raw file](https://github.com/cockroachdb/cockroach/blob/81a270e86c1c64052a1a0e5bab9598a4e83c7021/pkg/kv/kvserver/store_test.go#L2933)):*
> ```Go
> // prioritization of the snapshots as that is covered by the multi-queue
> // testing.
> func TestSendSnapshotConcurrency(t *testing.T) {
> ```

This test is certainly useful, though I think if at all possible it would be useful for us to have a test where we validate the entire snapshot flow (on sender/receiver) running concurrently w/ multiple senders.
___
*[`pkg/kv/kvserver/kvserverpb/raft.proto` line 207 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-NAkuvgME3COiiPUubbk:-NAkuvgME3COiiPUubbl:b-74hay2) ([raw file](https://github.com/cockroachdb/cockroach/blob/81a270e86c1c64052a1a0e5bab9598a4e83c7021/pkg/kv/kvserver/kvserverpb/raft.proto#L207)):*
> ```protobuf
> 
>     // The sending queue's name, to be utilized to ensure fairness across
>     // different snapshot sending sources. OTHER is a reserved name(0) and all
> ```

```
The default queue name, OTHER, is reserved for any uncategorized
and unprioritized snapshots, and requests with sender queue name OTHER
may not specify a non-zero sender_queue_priority.  To prioritize snapshots
categorized as OTHER, first...
```
___
*[`pkg/server/config.go` line 66 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-NAkvh4oEZ5mkb_9yDqf:-NAkvh4oEZ5mkb_9yDqg:b88qlzy) ([raw file](https://github.com/cockroachdb/cockroach/blob/81a270e86c1c64052a1a0e5bab9598a4e83c7021/pkg/server/config.go#L66)):*
> ```Go
> 	defaultScanMaxIdleTime    = 1 * time.Second
> 	defaultSnapshotSendLimit  = 2
> 	defaultSnapshotApplyLimit = 1
> ```

These are the ones I was thinking would be good to move.  I think you may have added a comment elsewhere, it might make sense to add a comment here that says
```
// ... See server.KVConfig for more info.
```
(or wherever else the comment lives).
___
*[`pkg/server/config.go` line 791 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-NAkw-PqDLCMFxWdBEZa:-NAkw-Pr8fT3H5Y_51vd:b-80u5io) ([raw file](https://github.com/cockroachdb/cockroach/blob/81a270e86c1c64052a1a0e5bab9598a4e83c7021/pkg/server/config.go#L791)):*
> ```Go
> 	cfg.ScanMaxIdleTime = envutil.EnvOrDefaultDuration("COCKROACH_SCAN_MAX_IDLE_TIME", cfg.ScanMaxIdleTime)
> 	cfg.SnapshotSendLimit = envutil.EnvOrDefaultInt64("COCKROACH_SNAPSHOT_SEND_LIMIT", cfg.SnapshotSendLimit)
> 	cfg.SnapshotApplyLimit = envutil.EnvOrDefaultInt64("COCKROACH_SNAPSHOT_APPLY_LIMIT", cfg.SnapshotApplyLimit)
> ```

Let's make sure these environment variables match the existing ones we had (I doubt anything relies on this, but I don't see much benefit in changing it).


<!-- Sent from Reviewable.io -->

AlexTalks reviewed (commented):
Per the testing discussion, in replica_learner_test.go something like the operations within `TestRaftSnapshotQueueSeesLearner` concurrently with some of the operations of `TestAddReplicaWithReceiverThrottling` or even `TestLearnerAdminChangeReplicasRace` could be useful - particularly if we can inject testing knobs for some snapshots to fail while others succeed.  `testRaftSnapshotsToNonVoters` seems particularly useful.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-:-NAlzr253E54J9wMtUZB:b-5tbw83)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aayushshah15)



<!-- Sent from Reviewable.io -->

Commit:
kvserver: honor priority on send and receive snapshots

Previously on both the send and receive snapshot side, the
various places that sent snapshots were uncoordinated, so
the choice of which snapshot was sent was somewhat arbitrary.
This PR uses the new multiqueue to prioritize them correctly.

Release note (performance improvement): Snapshots use a fair
round-robin approach for choosing which one to send next. This
allows decommissioning to complete much faster.

Release justification: A number of customer support cases
are caused by incorrect snapshot prioritization.

andrewbaptist reviewed (dismissed):
Working, looking at options for how to do this best. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-:-NAoCdHdAX1wKaQVyFSl:b-5s7ggh)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aayushshah15, @AlexTalks, and @nvanbenschoten)

___
*[`pkg/kv/kvserver/replica_command.go` line 2671 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-NAko2Hg4SKjS4ie8-HL:-NAo7lk12GFvqnimlLpB:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/81a270e86c1c64052a1a0e5bab9598a4e83c7021/pkg/kv/kvserver/replica_command.go#L2671)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

nit: Might be worth clarifying that the only thing we *for sure* want to remove in 23.1 is the cluster version.
</blockquote></details>

Done
___
*[`pkg/kv/kvserver/store.go` line 259 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-NAkoQaS3VvlvsLPcH9-:-NAo8h2TAghkq4E770kT:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/81a270e86c1c64052a1a0e5bab9598a4e83c7021/pkg/kv/kvserver/store.go#L259)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

I think (as we discussed) this should not use hard-coded values.  Apologies for making you keep moving these around, but maybe it would make sense to move the defaults into `kvserver` and that way you can reference them both here and in the `server` pkg.  You'll just have to make the defaults public however.
</blockquote></details>

Done
___
*[`pkg/kv/kvserver/store_snapshot.go` line 67 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-NAkthLnEaWOioiNdHJ4:-NAo8ilWE4SkwDPsSWOI:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/81a270e86c1c64052a1a0e5bab9598a4e83c7021/pkg/kv/kvserver/store_snapshot.go#L67)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

nit: `kv.snapshot_prioritization.enabled`
</blockquote></details>

Done
___
*[`pkg/kv/kvserver/store_snapshot.go` line 69 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-NAktlAh7OCMKyoEssnl:-NAo8jeyDZRt8xFNoXfw:b-vugxfa) ([raw file](https://github.com/cockroachdb/cockroach/blob/81a270e86c1c64052a1a0e5bab9598a4e83c7021/pkg/kv/kvserver/store_snapshot.go#L69)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

nit: does this pass lint?  I'd assume we'd need:
```
  true,
)
```
</blockquote></details>

Done (it did pass lint, but probably shouldn't have)
___
*[`pkg/kv/kvserver/store_snapshot.go` line 676 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-NAku1TZ5gr8yDIUv19d:-NAo8nLB8S5QOWxn-bJc:b7s74et) ([raw file](https://github.com/cockroachdb/cockroach/blob/81a270e86c1c64052a1a0e5bab9598a4e83c7021/pkg/kv/kvserver/store_snapshot.go#L676)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

nit: I still don't love the casting to `int`s, it feels like we should at least have had a `multiqueue.QueueID` type or something.  This isn't something I'd worry about changing now, but just wanted to note it.
</blockquote></details>

I don't love this either, but it does help in the case of version compatibility (it should definitely be strings over the wire, so it needs to cast to "something" - ints are just as good as strings from that perspective?
___
*[`pkg/kv/kvserver/store_snapshot.go` line 723 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-NAkuLHm8aSt77dREKAA:-NAo91EX2pwKD1msJIBD:b-m8drj3) ([raw file](https://github.com/cockroachdb/cockroach/blob/81a270e86c1c64052a1a0e5bab9598a4e83c7021/pkg/kv/kvserver/store_snapshot.go#L723)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

should we still call this `semaphore`?  It seems inaccurate - can we rename to something like `snapshotQueue`?
</blockquote></details>

Done, Renamed everywhere
___
*[`pkg/kv/kvserver/kvserverpb/raft.proto` line 207 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-NAkuvgME3COiiPUubbk:-NAo99XKEJiDOSm9l8ZQ:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/81a270e86c1c64052a1a0e5bab9598a4e83c7021/pkg/kv/kvserver/kvserverpb/raft.proto#L207)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

```
The default queue name, OTHER, is reserved for any uncategorized
and unprioritized snapshots, and requests with sender queue name OTHER
may not specify a non-zero sender_queue_priority.  To prioritize snapshots
categorized as OTHER, first...
```
</blockquote></details>

Done
___
*[`pkg/server/config.go` line 66 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-NAkvh4oEZ5mkb_9yDqf:-NAo9k4W4h9LkP4Z7bZJ:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/81a270e86c1c64052a1a0e5bab9598a4e83c7021/pkg/server/config.go#L66)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

These are the ones I was thinking would be good to move.  I think you may have added a comment elsewhere, it might make sense to add a comment here that says
```
// ... See server.KVConfig for more info.
```
(or wherever else the comment lives).
</blockquote></details>

Done
___
*[`pkg/server/config.go` line 791 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-NAkw-PqDLCMFxWdBEZa:-NAo9mOI-XLbFi2Y7JHn:bnhmw9f) ([raw file](https://github.com/cockroachdb/cockroach/blob/81a270e86c1c64052a1a0e5bab9598a4e83c7021/pkg/server/config.go#L791)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

Let's make sure these environment variables match the existing ones we had (I doubt anything relies on this, but I don't see much benefit in changing it).
</blockquote></details>

Done, I hadn't thought about the compatibility aspect here. 


<!-- Sent from Reviewable.io -->

AlexTalks reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86701)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aayushshah15, @herkolategan, and @nvanbenschoten)

___
*[`pkg/kv/kvserver/store_test.go` line 2933 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-NAkwNsXCu9MYfc8v2Qo:-NAz-1aJ6_-u1h6J1Ez6:b-n1jpk6) ([raw file](https://github.com/cockroachdb/cockroach/blob/81a270e86c1c64052a1a0e5bab9598a4e83c7021/pkg/kv/kvserver/store_test.go#L2933)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

This test is certainly useful, though I think if at all possible it would be useful for us to have a test where we validate the entire snapshot flow (on sender/receiver) running concurrently w/ multiple senders.
</blockquote></details>

Part of #87319


<!-- Sent from Reviewable.io -->

AlexTalks reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86701#-:-NAz-JP69nIR6dzdN2mJ:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @aayushshah15, @herkolategan, and @nvanbenschoten)



<!-- Sent from Reviewable.io -->

AlexTalks commented:
bors r+

AlexTalks commented:
bors r+
